File: frontend/src/store/friends.ts
import { create } from "zustand";
import { api } from "@/lib/api";
import type { BlockedUser, Friend, FriendListResponse, FriendRequest } from "@/lib/friends-types";

interface FriendsStore {
  friends: Friend[];
  pendingReceived: FriendRequest[];
  pendingSent: FriendRequest[];
  blocked: BlockedUser[];
  isLoading: boolean;
  fetch: () => Promise<void>;
  sendRequest: (userId: string) => Promise<void>;
  acceptRequest: (requestId: string) => Promise<void>;
  cancelRequest: (requestId: string) => Promise<void>;
  rejectRequest: (requestId: string) => Promise<void>;
  removeFriend: (userId: string) => Promise<void>;
  blockUser: (userId: string) => Promise<void>;
  unblockUser: (userId: string) => Promise<void>;
}

export const useFriendsStore = create<FriendsStore>((set, get) => ({
  friends: [],
  pendingReceived: [],
  pendingSent: [],
  blocked: [],
  isLoading: false,

  fetch: async () => {
    set({ isLoading: true });
    try {
      const { data } = await api.get<FriendListResponse>("/api/friends");
      set({
        friends: data.friends,
        pendingReceived: data.pending_received,
        pendingSent: data.pending_sent,
        blocked: data.blocked,
      });
    } catch {
      // ignore - user may not be logged in
    } finally {
      set({ isLoading: false });
    }
  },

  sendRequest: async (userId: string) => {
    await api.post(`/api/friends/request/${userId}`);
    await get().fetch();
  },

  acceptRequest: async (requestId: string) => {
    await api.post(`/api/friends/accept/${requestId}`);
    await get().fetch();
  },

  cancelRequest: async (requestId: string) => {
    set((state) => ({ pendingSent: state.pendingSent.filter((r) => r.id !== requestId) }));
    await api.delete(`/api/friends/request/${requestId}`);
    await get().fetch();
  },

  rejectRequest: async (requestId: string) => {
    await api.post(`/api/friends/reject/${requestId}`);
    await get().fetch();
  },

  removeFriend: async (userId: string) => {
    set((state) => ({ friends: state.friends.filter((f) => f.user_id !== userId) }));
    await api.delete(`/api/friends/${userId}`);
    await get().fetch();
  },

  blockUser: async (userId: string) => {
    await api.post(`/api/friends/block/${userId}`);
    await get().fetch();
  },

  unblockUser: async (userId: string) => {
    set((state) => ({ blocked: state.blocked.filter((u) => u.user_id !== userId) }));
    await api.post(`/api/friends/unblock/${userId}`);
    await get().fetch();
  },
}));

File: backend/app/api/routes/friends.py
"""Friend system API routes: /api/friends/*"""
import uuid

from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db
from app.schemas.friends import FriendListResponse, FriendRequestResponse
from app.services import auth_service
from app.services import friend_service

router = APIRouter()


async def _get_current_user(request: Request, db: AsyncSession):
    access_token = request.cookies.get("access_token")
    if not access_token:
        raise HTTPException(status_code=401, detail="Not authenticated")
    return await auth_service.get_current_user(db, access_token)


@router.get("", response_model=FriendListResponse)
async def list_friends(request: Request, db: AsyncSession = Depends(get_db)):
    """Get friends list, pending received, and pending sent requests."""
    user = await _get_current_user(request, db)
    return await friend_service.get_friends_list(db, user.id)


@router.post("/request/{user_id}", status_code=200)
async def send_request(user_id: uuid.UUID, request: Request, db: AsyncSession = Depends(get_db)):
    """Send a friend request to user_id. Always returns 200 to avoid browser console errors."""
    user = await _get_current_user(request, db)
    try:
        await friend_service.send_friend_request(db, user.id, user_id)
        return {"success": True}
    except HTTPException as e:
        return {"success": False, "detail": e.detail}


@router.post("/accept/{request_id}", response_model=FriendRequestResponse)
async def accept_request(request_id: uuid.UUID, request: Request, db: AsyncSession = Depends(get_db)):
    """Accept a pending friend request."""
    user = await _get_current_user(request, db)
    return await friend_service.accept_friend_request(db, request_id, user.id)


@router.delete("/request/{request_id}", status_code=204)
async def cancel_request(request_id: uuid.UUID, request: Request, db: AsyncSession = Depends(get_db)):
    """Cancel a pending friend request sent by the current user."""
    user = await _get_current_user(request, db)
    await friend_service.cancel_friend_request(db, request_id, user.id)


@router.post("/reject/{request_id}", response_model=FriendRequestResponse)
async def reject_request(request_id: uuid.UUID, request: Request, db: AsyncSession = Depends(get_db)):
    """Decline a pending friend request."""
    user = await _get_current_user(request, db)
    return await friend_service.reject_friend_request(db, request_id, user.id)


@router.delete("/{user_id}", status_code=204)
async def remove_friend(user_id: uuid.UUID, request: Request, db: AsyncSession = Depends(get_db)):
    """Remove an existing friend."""
    user = await _get_current_user(request, db)
    await friend_service.remove_friend(db, user.id, user_id)


@router.post("/block/{user_id}", status_code=204)
async def block_user(user_id: uuid.UUID, request: Request, db: AsyncSession = Depends(get_db)):
    """Block a user."""
    user = await _get_current_user(request, db)
    await friend_service.block_user(db, user.id, user_id)


@router.post("/unblock/{user_id}", status_code=204)
async def unblock_user(user_id: uuid.UUID, request: Request, db: AsyncSession = Depends(get_db)):
    """Unblock a previously blocked user."""
    user = await _get_current_user(request, db)
    await friend_service.unblock_user(db, user.id, user_id)
